File: leetcode/prob637.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

#include "gtest/gtest.h"

using namespace std;

// Name: 637. Average of Levels in Binary Tree
// Website: https://leetcode.com/problems/average-of-levels-in-binary-tree/
// Description:


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {

public:
    vector<double> averageOfLevels(TreeNode* root) {
        std::vector<double> results;
        if (!root) return results;
        std::queue<TreeNode *> q;
        q.push(root);
        int count = 0;
        int level_count = 1;
        int64_t level_sum = 0;
        while (!q.empty()) {
            TreeNode* cur = q.front(); q.pop();
            level_sum += cur->val;
            if (cur->left) {
                q.push(cur->left);
            }
            if (cur->right) {
                q.push(cur->right);
            }
            ++count;
            if (count == level_count) {
                count = 0;
                double avg = static_cast<double>(level_sum) / level_count;
                results.emplace_back(avg);
                level_sum = 0;
                level_count = q.size();
            }
        }
        return results;
    }

    vector<double> averageOfLevels_v2(TreeNode* root) {
        std::vector<double> results;
        if (!root) return results;
        std::queue<TreeNode *> q;
        q.push(root);
        while (!q.empty()) {
            size_t level_count = q.size();
            int64_t level_sum = 0;
            for (size_t i = 0; i < level_count; ++i) {
                TreeNode* cur = q.front(); q.pop();
                level_sum += cur->val;
                if (cur->left) {
                    q.push(cur->left);
                }
                if (cur->right) {
                    q.push(cur->right);
                }
            }
            results.emplace_back(static_cast<double>(level_sum)/level_count);
        }
        return results;
    }

};

Solution solu;

// TEST(Fuction, one) {}

File: leetCode_Python/prob637.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""

"""

# Name: 637. Average of Levels in Binary Tree
# Website: https://leetcode.com/problems/average-of-levels-in-binary-tree/
# Description:
import sys
from collections import defaultdict


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root):
        if root is None:
            return []
        levels = defaultdict(list)
        queue = [(root, 1)]
        level = 0
        while queue:
            node, level = queue.pop(0)
            levels[level].append(node.val)
            if node.left is not None:
                queue.append((node.left, level + 1))
            if node.right is not None:
                queue.append((node.right, level + 1))
        return [sum(levels[i]) / len(levels[i]) for i in range(1, level + 1)]


if __name__ == "__main__":
    pass